pkg/apis/core/types.go

```diff
@@ -4805,3 +4812,64 @@ const (
 	// DefaultHardPodAffinityWeight defines the weight of the implicit PreferredDuringScheduling affinity rule.
 	DefaultHardPodAffinitySymmetricWeight int32 = 1
 )
+
+type UnsatisfiableConstraintAction string
+
+const (
+	// DoNotSchedule instructs the scheduler not to schedule the pod
+	// when constraints are not satisfied.
+	DoNotSchedule UnsatisfiableConstraintAction = "DoNotSchedule"
+	// ScheduleAnyway instructs the scheduler to schedule the pod
+	// even if constraints are not satisfied.
+	ScheduleAnyway UnsatisfiableConstraintAction = "ScheduleAnyway"
+)
+
+// TopologySpreadConstraint specifies how to spread matching pods among the given topology.
+type TopologySpreadConstraint struct {
+	// MaxSkew describes the degree to which pods may be unevenly distributed.
+	// It's the maximum permitted difference between the number of matching pods in
+	// any two topology domains of a given topology type.
+	// For example, in a 3-zone cluster, MaxSkew is set to 1, and pods with the same
+	// labelSelector spread as 1/1/0:
+	// +-------+-------+-------+
+	// | zone1 | zone2 | zone3 |
+	// +-------+-------+-------+
+	// |   P   |   P   |       |
+	// +-------+-------+-------+
+	// - if MaxSkew is 1, incoming pod can only be scheduled to zone3 to become 1/1/1;
+	// scheduling it onto zone1(zone2) would make the ActualSkew(2-0) on zone1(zone2)
+	// violate MaxSkew(1).
+	// - if MaxSkew is 2, incoming pod can be scheduled onto any zone.
+	// It's a required field. Default value is 1 and 0 is not allowed.
+	MaxSkew int32
+	// TopologyKey is the key of node labels. Nodes that have a label with this key
+	// and identical values are considered to be in the same topology.
+	// We consider each <key, value> as a "bucket", and try to put balanced number
+	// of pods into each bucket.
+	// It's a required field.
+	TopologyKey string
+	// WhenUnsatisfiable indicates how to deal with a pod if it doesn't satisfy
+	// the spread constraint.
+	// - DoNotSchedule (default) tells the scheduler not to schedule it
+	// - ScheduleAnyway tells the scheduler to still schedule it
+	// It's considered as "Unsatisfiable" if and only if placing incoming pod on any
+	// topology violates "MaxSkew".
+	// For example, in a 3-zone cluster, MaxSkew is set to 1, and pods with the same
+	// labelSelector spread as 3/1/1:
+	// +-------+-------+-------+
+	// | zone1 | zone2 | zone3 |
+	// +-------+-------+-------+
+	// | P P P |   P   |   P   |
+	// +-------+-------+-------+
+	// If WhenUnsatisfiable is set to DoNotSchedule, incoming pod can only be scheduled
+	// to zone2(zone3) to become 3/2/1(3/1/2) as ActualSkew(2-1) on zone2(zone3) satisfies
+	// MaxSkew(1). In other words, the cluster can still be imbalanced, but scheduler
+	// won't make it *more* imbalanced.
+	// It's a required field.
+	WhenUnsatisfiable UnsatisfiableConstraintAction
```

Comment:
I'd still love to see some discussion in the comments of how this affects the pod *after* it's been scheduled. E.g. does it affect podDisruptionBudget? (I assume not at the moment but will it ever?) Descheduler? Scale-down behavior?

Reply:
@lavalamp  We could add more information on how it could affect later scheduling related decisions. Particularly, autoscaler simulates whether a pod being removed from a node can be scheduled on another node during scale-down. This feature may prevent a pod from being scheduled on a different node/zone/etc if the scheduling of the pod breaks the balanced spreading. Therefore, it could prevent scale-down in certain scenarios. This is actually a positive benefit. I have already heard from some users that the autoscaler sometimes aggressively removes nodes in clusters, causing a few pods of a service to be scheduled on a single node. This could reduce availability/reliability of a service if the node with a few pods of the service goes down.
This exact topic came up in our SIG Scheduling F2F meeting at KubeCon where people asked for a "required" spreading feature in order to prevent autoscaler from removing nodes.

All being said, I believe the right place for providing such information about this feature is in its documents. We cannot put all the information in doscstring of the API fields.

Reply:
Once a pod gets scheduled, default scheduler won't touch it any more. This is aligned with our design rationale, as re-scheduling may cause service down time, so we usually suggest users use external tool like descheduler to balance exsiting workloads proactively.

podDisruptionBudget is respected upon preemption, and that's covered in preeemption logic specifically.

Scale-down may be the real missing piece in this feature, at this moment we're relying on controller to take out excessive replicas, but it might not be optimal as controller doesn't have enough info to make that decision.

Reply:
I think my main point here is that people reading these api comments will not only care about the scheduler, they will care about the entire lifecycle of their pods / services. I think it's reasonable to at least give a breadcrumb here to other documentation to learn about how the system behaves after these pods have been scheduled.

Reply:
It makes sense. I will include a link here once the doc is baked.

Reply:
When we say scale down are we talking about using the scale subresource to reduce the replica count of a Deployment or otherwe resources ?For that we would need logic in the deployment controller to say bring down a pod , which doesnt affect even spreading

Reply:
@krmayankk you're right.